File: 0017_letter_combinations_of_a_phone_numer/main.py
class Solution:
    res = []
    mapping = {
        "2": ["a", "b", "c"],
		"3": ["d", "e", "f"],
		"4": ["g", "h", "i"],
		"5": ["j", "k", "l"],
		"6": ["m", "n", "o"],
		"7": ["p", "q", "r", "s"],
		"8": ["t", "u", "v"],
		"9": ["w", "x", "y", "z"],
    }

    def letterCombinations(self, digits):
        """
        :type digits: str
        :rtype: List[str]
        """
        self.res = []
        if len(digits) != 0:
            self.findCombinations(0, digits, "")
        return self.res
        
    def findCombinations(self, pos, digits, comb):
        if pos >= len(digits):
            self.res.append(comb)
            return
        curr = self.mapping.get(digits[pos])
        if curr is not None:
            for c in curr:
                self.findCombinations(pos+1, digits, comb+c)

File: 0017_letter_combinations_of_a_phone_numer/main.go
package main

var (
	mapping = map[byte][]string{
		'2': []string{"a", "b", "c"},
		'3': []string{"d", "e", "f"},
		'4': []string{"g", "h", "i"},
		'5': []string{"j", "k", "l"},
		'6': []string{"m", "n", "o"},
		'7': []string{"p", "q", "r", "s"},
		'8': []string{"t", "u", "v"},
		'9': []string{"w", "x", "y", "z"},
	}
)

func findCombination(res *[]string, pos int, digits string, comb string) {
	if pos >= len(digits) {
		tmp := make([]byte, len(comb))
		copy(tmp, comb)
		*res = append(*res, string(comb))
		return
	}
	curr, ok := mapping[digits[pos]]
	if ok {
		for _, c := range curr {
			findCombination(res, pos+1, digits, comb+c)
		}
	}
}

func letterCombinations(digits string) []string {
	res := []string{}
	if len(digits) != 0 {
		findCombination(&res, 0, digits, "")
	}
	return res
}

func main() {

}
